honeybee_grasshopper_energy/src/HB Opaque Construction.py

```diff
@@ -0,0 +1,86 @@
+# Honeybee: A Plugin for Environmental Analysis (GPL)
+# This file is part of Honeybee.
+#
+# Copyright (c) 2019, Ladybug Tools.
+# You should have received a copy of the GNU General Public License
+# along with Honeybee; If not, see <http://www.gnu.org/licenses/>.
+# 
+# @license GPL-3.0+ <http://spdx.org/licenses/GPL-3.0+>
+
+"""
+Create an EnergyPlus opaque construction. Layer inputs can be either the name of
+a material within the material library or a custom material made with any of the
+EnergyPlus Material components.
+_
+Note that _layer_1 is always the outermost (exterior) layer and the last layer
+in the component is always the innermost (interior) layer.
+_
+To add more layers in the construction, zoom into the component and hit the
+lowest "+" sign that shows up on the input side. 
+_
+To remove layers from the construction, zoom into the component and hit the
+lowest "-" sign that shows up on the input side.
+-
+
+    Args:
+        _name: A unique name for the opaque construction.
+        _layer_1: The first outer-most material layer of the construction.
+        _layer_2: The second outer-most material layer of the construction.
+        _layer_3: The third outer-most material layer of the construction.
+        _layer_4: The fourth outer-most material layer of the construction.
+        _layer_5: The fifth outer-most material layer of the construction.
+        _layer_6: The sixth outer-most material layer of the construction.
+        _layer_7: The seventh outer-most material layer of the construction.
+        _layer_8: The eigth outer-most material layer of the construction.
+        _layer_9: The ninth outer-most material layer of the construction.
+        _layer_10: The tenth outer-most material layer of the construction.
```

Comment:
Are we exposing the maximum 10 layers? I see that inside the code you are taking care of automatic input assignment which probably means the answer is no.

I have been thinking for some time that we should just have a single input for layers and make it list access. Taking care of order of layers is very basic in Grasshopper and people should be able to do that.

Keeping the inputs / outputs static makes the whole process of translating the components to other formats (Dynamo, Web, ...) much easier. It will also make the process of updating the definition easier and most straight forward.

Reply:
You're right that I'm only exposing 3 by default and, because I auto-generate the input description, there's really no need for me to put all of this here (it will just make the docs generated from the components harder to read).  I'll change it to only have 3 layers here.

Reply:
While it's true that experienced users should be able to control the order of a list in Grasshopper, I don't think any anyone would deny that it is more convenient and human-readable when we use dynamic inputs.  It definitely saves me from having to check the order of the materials in the output `__str__` representation of the construction, which I would feel I have to do it if were a list input.  Dynamic inputs also makes it easier to read the input docstrings to remember which layer is outermost vs. innermost (I still forget whether it's the first or last layer).

Also, I feel that your concerns about translating to Dynamo or Web are addressed by the fact that the Python API uses a list input on the Construction objects. So, it's relatively easy to make a graphic interface that uses a list if that is truly what is best for that interface.  But I don't think a list is the best type of input for the Revit interface that Konrad is developing right now, which is currently the only other GUI we are working on at the moment.  For that reason, I also think it might not be the best for a web interface, though that can depend on other factors that are hard to predict.

So that just leaves Dynamo and, at this point, we have already made a lot of decisions that break the direct translation of Grasshopper components to Dynamo ones (the use of `isinstance` instead of `hasattr` being a major one). And I think that we are generally in agreement that we shouldn't make our libraries or other interfaces suffer just to accommodate Dynamo.

Reply:
I think my extra explanation was misleading. The point is let's avoid components with dynamic inputs and outputs unless there is no other way to do it. I don't think that's the case here. If one is concerned about the order they can use the `merge` component in Grasshopper.

Reply:
Maybe we should set up a call since I'm still unclear on why we should do this.

Reply:
After a call, I understand Mostapha's concern here better and I will switch over the two dynamic components to be static.

Reply:
Done!